Add IS NOT? DISTINCT FROM predicate

marstein opened the pull request:
Added sql parsing and execution classes for the DISTINCT predicate.
Closes #14644

## Summary of the changes / Why this improves CrateDB
Implement the IS DISTINCT FROM predicate. It is just like `<>` but handles NULL values differently.
TODO: arrays and possibly type check.
The predicate is implemented as a Function. Please review if this is the right choice.

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x ] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

crate-jenkins commented:
Can one of the admins verify this patch?

matriv commented:
ok to test

mfussenegger commented:
> The predicate is implemented as a Function. Please review if this is the right choice.

Yes, implementing it internally as a scalar sounds good.

marstein commented on server/src/testFixtures/java/io/crate/expression/scalar/ScalarTestCase.java:
```diff
@@ -23,7 +23,6 @@
 
 import static io.crate.testing.Asserts.assertThat;
 import static io.crate.testing.Asserts.isNull;
-import static org.assertj.core.api.Assertions.assertThat;
```
I am not sure how and why I removed this. It looks as if the io.crate `assertThat` extends the assertj version and thus the seconds `import static` is not necessary.

marstein commented:
Hello, @mfussenegger or @matriv - could you please review before I forget it all. Thanks!

jeeminso commented on libs/sql-parser/src/main/java/io/crate/sql/ExpressionFormatter.java:
```diff
@@ -675,6 +676,12 @@ protected String visitBetweenPredicate(BetweenPredicate node, @Nullable List<Exp
                    node.getMax().accept(this, parameters) + ")";
         }
 
+        @Override
+        protected String visitDistinctFrom(DistinctFromPredicate node, @Nullable List<Expression> parameters) {
+            return "(" + node.getLeft().accept(this, parameters) + " IS DISTINCT FROM " +
+                   node.getRight().accept(this, parameters) + ")";
+        }
+
```
We could re-use `visitComparisonExpression`

jeeminso commented on server/src/main/java/io/crate/expression/operator/DistinctFromPredicate.java:
```diff
@@ -0,0 +1,110 @@
+        if (allArgsAreNull(function)) {
+            return Literal.of(RETURN_TYPE, false);
+        }
+        if (oneArgIsNull(function) && bothArgsAreLiteral(function)) {
+            return Literal.of(RETURN_TYPE, true);
+        }
```
I think `bothArgsAreLiteral` is not needed, `evaluateIfLiterals` should be handling it. Could you double check?

jeeminso commented on server/src/main/java/io/crate/expression/operator/DistinctFromPredicate.java:
```diff
@@ -0,0 +1,110 @@
+        }
+    }
+
+    private static boolean allArgsAreNull(Function function) {
+        return function.arguments().stream().allMatch(arg -> arg instanceof Input<?> && ((Input<?>) arg).value() == null);
+    }
+
+    private static boolean oneArgIsNull(Function function) {
+        return function.arguments().stream().anyMatch(arg -> arg instanceof Input<?> && ((Input<?>) arg).value() == null);
+    }
```
minor: there is a slight redundancy, we could simplify this so it would look more like the `evaluate` function.

jeeminso reviewed (commented):
Thanks a lot for you contribution @marstein !

marstein commented:
@jeeminso Feedback applied! Saved many lines of duplicate code. 
The `DistinctPredicate.NAME` now contains spaces, I hope that is intended and correct.

jeeminso commented on server/src/test/java/io/crate/expression/predicate/DistinctFromPredicateTest.java:
```diff
@@ -0,0 +1,296 @@
+A <distinct predicate> tests two values to see whether they are distinct and
+returns either ``TRUE`` or ``FALSE``.
+ */
+public class DistinctFromPredicateTest extends ScalarTestCase {
```
`IS NOT DISTINCT` tests do not seem necessary, `NotPredicate` should be able to handle any kind of child expressions.

Similarly, I am not sure if the array variants are necessary but I can see that you referenced `EqOperatorTest`.

mfussenegger replied:
I agree that some of these tests seem a bit redundant as the cases are covered by the `DataType.compare` implementation/tests

I'd reduce it as follows:
- Keep the ones specific to the null handling of distinct from
- Merge the `<value> is (not) distinct from <value>` cases into a single case that uses `DataTypeTesting.randomType()` + `DataTypeTesting.getDataGenerator`

jeeminso reviewed (commented):
Thank you for your work again @marstein , it looks good overall but I personally find it a bit too much tests. It would be good to wait for others inputs as well.

mfussenegger commented on server/src/main/java/io/crate/expression/predicate/Predicates.java:
```diff
@@ -36,5 +37,6 @@ public void addFunctions(Settings settings,
         IsNullPredicate.register(builder);
         NotPredicate.register(builder);
         MatchPredicate.register(builder);
+        DistinctFromPredicate.register(builder);
```
```suggestion
```

Only one registration is necessary and there is already one in `Operators`

mfussenegger commented on server/src/main/java/io/crate/expression/operator/DistinctFromPredicate.java:
```diff
@@ -0,0 +1,90 @@
+import io.crate.types.DataType;
+import io.crate.types.TypeSignature;
+
+public class DistinctFromPredicate<T> extends Operator<Object> {
```
```suggestion
public class DistinctFrom<T> extends Operator<Object> {
```

mfussenegger commented on server/src/main/java/io/crate/expression/operator/DistinctFromPredicate.java:
```diff
@@ -0,0 +1,90 @@
+        Object arg2 = args[1].value();
+
+        // two ``NULL`` values are not distinct from one other
+        if (arg1 == null && arg2 == null)
+            return false;
+
+        // Any non-null Literal is distinct from null
+        if (arg1 == null || arg2 == null)
+            return true;
+
```
```suggestion
        if (arg1 == null && arg2 == null) {
            return false;
        }
        // Any non-null Literal is distinct from null
        if (arg1 == null || arg2 == null) {
            return true;
        }
```

mfussenegger commented on server/src/test/java/io/crate/integrationtests/DistinctFromIntegrationTest.java:
```diff
@@ -0,0 +1,82 @@
+import org.junit.Test;
+
+@IntegTestCase.ClusterScope(minNumDataNodes = 2)
+public class DistinctFromIntegrationTest extends IntegTestCase {
```
I think these could be merged together with the `DistinctFromQueryTest` into a test suite that uses the `QueryTester` to do both query type assertion and result-matching assertions. 
The `EqOperatorTest` has examples for this like `test_terms_query_on_empty_object`

marstein replied:
Aha, that is how you do that... I added a new `io.crate.expression.predicate.DistinctFromTest.test_terms_query_on_empty_object`.

marstein replied:
I removed the integration tests

mfussenegger commented on server/src/test/java/io/crate/expression/predicate/DistinctFromPredicateTest.java:
```diff
@@ -0,0 +1,296 @@
+    //  must be of the same degree and each corresponding pair of Fields must have
+    //  comparable <data type>s.
+    //
+    @Test(expected = ClassCastException.class)
+    public void testEvaluateIncomparableDatatypes() {
+        assertEvaluate("3 is distinct from true", isLiteral(false));
+    }
```
```suggestion
    @Test
    public void testEvaluateIncomparableDatatypes() {
        assertThatThrownBy(() -> assertEvaluate("3 is distinct from true", isLiteral(false)))
            .isExactlyInstanceOf(ClassCastException.class);
    }
```

mfussenegger commented on server/src/test/java/io/crate/expression/predicate/DistinctFromPredicateTest.java:
```diff
@@ -0,0 +1,296 @@
+    //  must be of the same degree and each corresponding pair of Fields must have
+    //  comparable <data type>s.
+    //
+    @Test(expected = java.lang.ClassCastException.class)
+    public void testNotEvaluateIncomparableDatatypes() {
+        assertEvaluate("3 is not distinct from false", isLiteral(false));
+    }
```
```suggestion
    @Test
    public void testNotEvaluateIncomparableDatatypes() {
        assertThatThrownBy(() -> assertEvaluate("3 is not distinct from false", isLiteral(false)))
            .isExactlyInstanceOf(ClassCastException.class);
    }
```

mfussenegger reviewed (commented):
Thanks for the great work.

I left a couple more comments, some of which are mostly minor style nitpicks. (I'll follow up and tune the checkstyle rules as I was surprised it didn't catch some of it)

I also think we should have some more docs for this. E.g. under `docs/general/dql/selects.rst` in the comparison section - with a huge warning that the operator won't use index structures and is (currently) unsuitable for use on larger tables.

marstein commented:
I addressed all of the feedback, but moving the comparison tests into some kind of automated mechanism baffles me: it is not possible to compare random types, as some are not meant to be compared. So only comparable types should be tested. How can this be achieved with random types? Randomness and testing are a bit of a stretch anyhow...

mfussenegger commented on docs/general/dql/selects.rst:
```diff
@@ -509,6 +509,49 @@ does always return ``NULL`` when comparing ``NULL``.
+Returns ``TRUE`` if ``expr`` is not equal to :ref:`evaluate <gloss-evaluation>`. If both expressions are ``NULL`` the result is false. Comparing a non-NULL value to ``NULL``
+returns ``TRUE`` - ``NULL`` is considered distinct from any value.
+
+Use this predicate to check for non-``NULL`` values as SQL's three-valued logic
+does always return ``NULL`` when comparing ``NULL``.
+
+.. vale off
+
+:expr:
+  :ref:`Expression <gloss-expression>` of one of the supported
+  :ref:`data types <data-types>` supported by CrateDB.
+
+.. vale on
+
+::
+
+    cr> select name from locations where inhabitants['interests'] is distinct from 'Argabuthon'
+    ... order by name;
+    +-------------------+
+    | name              |
+    +-------------------+
+    | Arkintoofle Minor |
+    | Bartledan         |
+    +-------------------+
+    SELECT 2 rows in set (... sec)
+
+::
+
+    cr> select count(*) from locations where name is distinct from 'Argabuthon';
+    +----------+
+    | count(*) |
+    +----------+
+    |       12 |
+    +----------+
+    SELECT 1 row in set (... sec)
+
+.. NOTE::
+
+   Using ``IS DISTINCT FROM`` can be slow because the expression is currently not using index structures and is not suitable for use on larger tables.
+
```
```suggestion
``IS DISTINCT FROM``
____________________

Returns ``TRUE`` if ``expr`` is not equal to :ref:`evaluate <gloss-evaluation>`.
If both expressions are ``NULL`` the result is false. Comparing a non-NULL value
to ``NULL`` returns ``TRUE`` - ``NULL`` is considered distinct from any value.

Use this predicate to check for non-``NULL`` values as SQL's three-valued logic
does always return ``NULL`` when comparing ``NULL``.

.. vale off

:expr:
  :ref:`Expression <gloss-expression>` of one of the supported
  :ref:`data types <data-types>` supported by CrateDB.

.. vale on

::

    cr> SELECT
    ...   name,
    ...   name IS DISTINCT FROM 'Trillian' as is_distinct,
    ...   name != 'Trillian' as not_eq
    ... FROM unnest(['Arthur', 'Trillian', null]) as t(name)
    ... ORDER BY 1;
    +----------+-------------+--------+
    | name     | is_distinct | not_eq |
    +----------+-------------+--------+
    | Arthur   | TRUE        | TRUE   |
    | Trillian | FALSE       | FALSE  |
    | NULL     | TRUE        | NULL   |
    +----------+-------------+--------+
    SELECT 3 rows in set (... sec)


.. NOTE::

   Using ``IS DISTINCT FROM`` can be slow because the expression is not using
   index structures and is not suitable for use on larger tables.
```


Adapted the formatting to pass checks and changed the example to highlight the null/3-vl difference

mfussenegger reviewed (commented):
>  but moving the comparison tests into some kind of automated mechanism baffles me: it is not possible to compare random types, as some are not meant to be compared. So only comparable types should be tested. How can this be achieved with random types? Randomness and testing are a bit of a stretch anyhow...

See my suggestion. It can use the same value twice for once test, and generate a second value that's different for another.

This acts a bit like a property test where using random value generation is common.

marstein commented on server/src/test/java/io/crate/expression/predicate/DistinctFromTest.java:
```diff
@@ -0,0 +1,233 @@
+        assertNormalize("[0.0012205839] IS DISTINCT FROM 32569", isLiteral(Boolean.TRUE));
+    }
+
+    //    @Ignore("this fails for multiple combinations; e.g. UnsupportedFunction Unknown function: (true IS DISTINCT FROM B'0'), no overload found for matching argument types: (boolean, bit)")
+    // io.crate.exceptions.ConversionException: Cannot cast `'Pt(x=-117.85833853141307,y=89.9999999999999)'` of type `text` to type `double precision`
+    // io.crate.exceptions.ConversionException: Cannot cast value `191.91.237.213` to type `object`
+    // io.crate.exceptions.UnsupportedFunctionException: Unknown function: ([0.0351398] IS DISTINCT FROM _map('x', -5641633869089706770::bigint)), no overload found for matching argument types: (double precision_array, object). for [0.0351398] IS DISTINCT FROM {"x"=-5641633869089706770::bigint}
+    // io.crate.exceptions.UnsupportedFunctionException: Unknown function: ('maHzNLMRF' IS DISTINCT FROM [0.72332287]), no overload found for matching argument types: (text, double precision_array).
+    @Test
+    public void test_random_data_type() {
+        DataType<?> randomType = randomTypeExcluding(Set.of(DataTypes.BOOLEAN));
+        // Random value IS NOT distinct from itself.
+        Supplier<?> dataGenerator = getDataGenerator(randomType);
+        Literal<?> val = Literal.ofUnchecked(randomType, dataGenerator.get());
+        assertEvaluate(val.toString(Style.QUALIFIED) + " IS NOT DISTINCT FROM " + val.toString(Style.QUALIFIED), Boolean.TRUE);
+
+        // Random value IS DISTINCT from another random value.
+        Literal<?> anotherVal = Literal.ofUnchecked(randomType, dataGenerator.get());
+        assertEvaluate(val.toString(Style.QUALIFIED) + " IS DISTINCT FROM " + anotherVal.toString(Style.QUALIFIED), Boolean.TRUE);
+
+        // Two random values of different types are NOT DISTINCT from each other.
+        DataType<?> randomType2;
+        do {
+            randomType2 = randomType();
+        } while (randomType2.equals(randomType));
+        Literal<?> valDiffType = Literal.ofUnchecked(randomType2, getDataGenerator(randomType2).get());
+        assertEvaluate(val.toString(Style.QUALIFIED) + " IS DISTINCT FROM " + valDiffType.toString(Style.QUALIFIED), Boolean.TRUE);
+    }
```
This is great. I am still working on understanding the DataType/TypeSignature system. The "parameterized types" in the assertEvaluate are cool. Maybe assertEvaluate could in the future return the XXAssert Object so one could put a comment on the assert, like `.as("non-nuill value compared to null should be distinct")` -- just a suggestion

marstein commented:
Everything is ready now. Let me know - I just need to squash unless Github does this for us.

jeeminso commented on server/src/test/java/io/crate/expression/predicate/DistinctFromTest.java:
```diff
@@ -0,0 +1,216 @@
+        }
+    }
+
+    // --------------------------------------------------------------------------------------------
+    // Array comparison: For arrays,
+    // ``IS DISTINCT FROM`` tests each corresponding pair of array elements. If the
+    // arrays are empty, or if any element of the first array contains the same value
+    // as its corresponding element in the second array, or if both elements are
+    // ``NULL``, ``IS DISTINCT FROM`` returns ``FALSE``.
```
I think `test_random_data_type` covers the array variants below these lines.

jeeminso commented on server/src/test/java/io/crate/expression/predicate/DistinctFromTest.java:
```diff
@@ -0,0 +1,216 @@
+    public void test_random_data_type_against_null() {
+        DataType<?> randomType = randomType();
+        var val = Literal.ofUnchecked(randomType, getDataGenerator(randomType).get()).toString(Style.QUALIFIED);
+        // Random value IS NOT distinct from null.
+        assertEvaluate("? IS NOT DISTINCT FROM null", Boolean.FALSE, Literal.of(val));
+        assertEvaluate("null IS NOT DISTINCT FROM ?", Boolean.FALSE, Literal.of(val));
```
Sorry for nitpicking, `IS NOT DISTINCT` is basically `NOT(IS DISTINCT)` so the below two asserts are enough.

marstein replied:
Sure, I removed those. I just would not assume that NOT does always get properly evaluated from experience...

jeeminso reviewed (commented):
Left a few more minor comments, please squash the commits, Thank you!

marstein commented:
Alright - changes applied - extra tests removed, squashed.

jeeminso commented on docs/general/dql/selects.rst:
```diff
@@ -509,6 +509,46 @@ does always return ``NULL`` when comparing ``NULL``.
+.. NOTE::
+
+   Using ``IS DISTINCT FROM`` can be slow because the expression is not using
+   index structures and is not suitable for use on larger tables.
 .. _sql_dql_array_comparisons:
```
```suggestion
   index structures and is not suitable for use on larger tables.
   
   
.. _sql_dql_array_comparisons:
```

jeeminso commented on docs/general/dql/selects.rst:
```diff
@@ -509,6 +509,46 @@ does always return ``NULL`` when comparing ``NULL``.
    <scalar-null-or-empty>` scalar for improved performance.
 
 
+``IS DISTINCT FROM``
+____________________
```
```suggestion
.. _sql_dql_is_distinct_from:

``IS DISTINCT FROM``
--------------------
```

jeeminso reviewed (approved):
Thanks a lot for your work! I think there are few style changes required to fix the sphinx failures which I am not too familiar either. Could you try the suggestions?

marstein commented:
I applied your changes and ran `sphinx` locally, including the tests, in blackbox/bin and everything comes out clean looking and tests pass. I think you fixed it.
